File: Server_Code/Packet/LoginSuccessAckPacket.cpp
#include "LoginSuccessAckPacket.h"

#include <stdexcept>

LoginSuccessAckPacket::LoginSuccessAckPacket(uint32_t uid, const std::string& nickname)
    : LoginPacket(EPacketType::LOGIN_SUCCESS_ACK, nickname), uid(uid)
{
    header.size = static_cast<uint16_t>(
        sizeof(PacketHeader) + sizeof(uint32_t) + nickname.length() + 1
        );
}

LoginSuccessAckPacket::LoginSuccessAckPacket()
    : LoginPacket(EPacketType::LOGIN_SUCCESS_ACK, ""), uid(0)
{
}

std::vector<char> LoginSuccessAckPacket::Serialize() const
{
    BinaryWriter writer;
    writer.WriteHeader(header);
    writer.WriteUInt32(uid);
    writer.WriteString(nickname);
    return writer.GetBuffer();
}

void LoginSuccessAckPacket::Deserialize(const char* buffer, size_t bufferSize)
{
    BinaryReader reader(buffer, bufferSize);
    header = reader.ReadHeader();
    uid = reader.ReadUInt32();
    nickname = reader.ReadString();
}

File: Assets/Script/Packet/LoginSuccessAckPacket.cs
using System;
using System.Text;
using NetworkUtil;

/// <summary>
/// 서버가 클라이언트에게 로그인 성공 사실과 UID, 닉네임을 전달하는 패킷
/// </summary>
public class LoginSuccessAckPacket : LoginPacket
{
    public override EPacketType PacketType => EPacketType.LOGIN_SUCCESS_ACK;

    public uint UID { get; set; }

    public LoginSuccessAckPacket() : base("") { }

    public LoginSuccessAckPacket(uint uid, string nickname) : base(nickname)
    {
        UID = uid;
    }

    public override void Serialize(BinaryWriter writer)
    {
        writer.WriteUInt32(UID);
        writer.WriteString(Nickname);
    }

    public override void Deserialize(BinaryReader reader)
    {
        UID = reader.ReadUInt32();
        Nickname = reader.ReadString();
    }

    public override bool IsValid()
    {
        return UID > 0 && !string.IsNullOrWhiteSpace(Nickname) && Encoding.UTF8.GetByteCount(Nickname) <= 255;
    }
}
